src/language/patterns.md

```diff
@@ -0,0 +1,358 @@
+[variable patterns][variable] or [constant patterns][constant]:
+
+```dart
+switch (obj) {
+  // List pattern [a,b] matches obj first if obj is a list wth two fields,
+  // then if its fields match the constant subpatterns 'a' and 'b'.
+  case [a, b]: // ...   
+}                    
+```
+
+Patterns can ignore parts of a matched value by using a [wildcard pattern][]
+as a place holder, or a [rest element][] in the case of list patterns.
+
+### Destructuring
+
+When an object and pattern match, the pattern can then access the object's data 
+and extract it in parts. In other words, the pattern _destructures_ the object:
+
+```dart
+var numList = [1, 2, 3];
+// List pattern [a, b, c] destructures the three elements from numList...
+var [a, b, c] = numList;
+// ...and assigns them to new variables.
+print(a + b + c);          
+```
+
+You can nest [any kind of pattern][types] inside a destructuring pattern, 
+For example, this case pattern matches and destructures a two-element
+list whose first element is `a` or `b`:
+
+```dart
+switch (list) {
+  case ['a' || 'b', var c]: 
+  // ...
+}
+```
+
+## Places patterns can appear
+
+You can use patterns in several places in the Dart language:
+
+<a id="pattern-uses"></a>
+
+- Local variable [declarations](#variable-declaration) and [assignments](#variable-assignment)
+- [for and for-in loops][for]
+- [if-case][if] and [switch-case][switch]
+- Control flow in [collection literals][]
+
+This section describes common use cases for matching and destructuring with patterns, 
+combining various contexts, pattern types, and surrounding code constructs.
+
+### Variable declaration
+
+Anywhere Dart allows local variable declaration, like in the body of a function,
+you can also use a _pattern variable declaration_. The pattern matches against
+and destructures the value on the right of the declaration, binding it to
+new local variables:
+
+```dart
+// Declares new variables a, b, and c.
+var (a, [b, c]) = ("str", [1, 2]);
+```
+
+A pattern variable declaration must start with either `var` or `final`, followed
+by a pattern. 
+
+### Variable assignment 
+
+A _variable assignment pattern_ on the left side of an assignment destructures the
+matched object and assigns the values to _existing_ variables,
+instead of binding new ones. 
+
+One convenient use case is swapping the values of two variables without needing
+to declare a third temporary one:
+
+```dart
+var (a, b) = ('left', 'right');
+(b, a) = (a, b); // Swap.
+print('$a $b'); // Prints "right left".
+```
+
+### Switch statements and expressions
+
+Every case clause contains a pattern. This applies to [switch statements][switch]
+and [expressions][], as well as [if-case statements][if].
+
+_Case patterns_ are [refutable][];
+they allow control flow to either match and destructure the object being switched
+on, or continue execution if the object doesn't match. Case patterns can be
+[any kind of pattern][types]:
```

Comment:
**issue:** If you can use a semi-colon, you can use a period.
```suggestion
_Case patterns_ are [refutable][].
They control flow toward one of two outcomes:
* Match and destructure the switched object.
* Continue execution if the object doesn't match.
Case patterns can be [any kind of pattern][types].
```

**nit:** I would change the first sentence to active voice, but that would involve changing the link.

Reply:
I'm not sure how the first sentence can be made active. Ignoring the link, what would you do?

Reply:
**comment:** What refutes case patterns? That would explain it.

That said, I'm trying to soften my approach to passive voice. The **When do I use passive voice?** section of [this page](https://advice.writing.utoronto.ca/revising/passive-voice/) from the University of Toronto seems to be a good guide. Look at that to see if what you have here just makes more sense to keep as is.

Reply:
Oh, ok. It didn't even cross my mind that that kind of sentence structure (Case patterns are refutable) is passive voice, good to know!

Who or what is refuting them is really hard to say so I'm going to leave this for now. I think it might be the pattern that does the refuting, but also is the one being refuted...idk. I vaguely remember Bob saying "refutable/irrefutable" are _types_ of patterns, so the passive voice is built into the descriptors.